### PR TITLE
Remove type filter from GetMappings API

### PR DESCRIPTION
**Pull request by romseygeek**
This commit removes the types filter from the GetMappings API, which is no longer
useful seeing as we can only have a single mapping type per index.  It also changes
the structure of GetMappingsResponse and GetIndexResponse to remove the extra
nesting of mappings below the no-longer-relevant type name, and removes the types
methods from the equivalent request classes.

Relates to #41059

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateServiceTests.java [hunk 1]
```diff
@@ -333,7 +333,7 @@ public void testAddIndexClosedBlocksReusesBlocks() {
     }
 
     public void testValidateShardLimit() {
-        int nodesInCluster = randomIntBetween(2,100);
+        int nodesInCluster = randomIntBetween(2,90);
```
Comment (colings86) [inline server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateServiceTests.java]: This change seems unrelated to types removal?

Comment (romseygeek) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateServiceTests.java]: Fixed by a separate merge in from master

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetMappingsResponseTests.java [hunk 2]
```diff
@@ -19,54 +19,40 @@
...
     }
 
-    private static void assertEqualInstances(GetMappingsResponse expected, GetMappingsResponse actual) {
-        assertEquals(expected.mappings(), actual.mappings());
+    @Override
+    protected GetMappingsResponse doParseToClientInstance(XContentParser parser) throws IOException {
+        return GetMappingsResponse.fromXContent(parser);
     }
 
-    private Predicate<String> randomFieldsExcludeFilter() {
```
Comment (jtibshirani) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetMappingsResponseTests.java]: It could make sense to keep this as a precaution, to ensure the mappings are always well-formed.

**Review by jtibshirani — COMMENTED** [review]
This looks good to me overall, I left a few minor comments.

I was a bit confused at first about what this PR did -- we're removing deprecated HLRC methods and also refactoring transport client classes at the same time. For future APIs perhaps we could split these into two separate tasks, or just clarify in the PR description? It could also be helpful to mention 'transport API' in the PR title, otherwise it's easy to assume it refers to the REST API.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetMappingsResponseTests.java (hunk 2 above)
Comment (romseygeek) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetMappingsResponseTests.java]: HLRC's AbstractResponseTestCase doesn't seem to support this, unfortunately

**Comment by romseygeek** [comment]
> I was a bit confused at first about what this PR did -- we're removing deprecated HLRC methods and also refactoring transport client classes at the same time. For future APIs perhaps we could split these into two separate tasks, or just clarify in the PR description? It could also be helpful to mention 'transport API' in the PR title, otherwise it's easy to assume it refers to the REST API.

Yes, sorry about that.  I think I'll open a separate PR to just remove all the type-referring deprecated methods in HLRC to prevent confusion in the future

File: server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java [hunk 3]
```diff
@@ -42,15 +42,19 @@ public ClusterInfoRequest() {
     public ClusterInfoRequest(StreamInput in) throws IOException {
         super(in);
         indices = in.readStringArray();
-        types = in.readStringArray();
+        if (in.getVersion().before(Version.V_8_0_0)) {
+            in.readStringArray();
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java]: I think these can still be non-empty in a mixed cluster with a user-request containing types routed through a 7x node

File: server/src/main/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponse.java [hunk 4]
```diff
@@ -29,117 +30,87 @@
...
+            if (in.getVersion().before(Version.V_8_0_0)) {
+                int mappingCount = in.readVInt();
+                assert mappingCount == 1 || mappingCount == 0 : "Expected 0 or 1 mappings but got " + mappingCount;
+                if (mappingCount == 1) {
+                    String type = in.readString();
+                    assert MapperService.SINGLE_MAPPING_NAME.equals(type) : "Expected type [_doc] but got [" + type + "]";
+                    indexMapBuilder.put(index, new MappingMetaData(in));
+                }
+                else {
+                    indexMapBuilder.put(index, null);
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponse.java]: I agree, `null` is nasty here.  I've pushed a change to create a new empty mappings placeholder.

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/2
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

File: server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java]: I don't think we support the case where a user is making typed requests in a mixed 7.x and 8.x cluster?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetMappingsResponseTests.java (hunk 2 above)
Comment (jtibshirani) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/indices/GetMappingsResponseTests.java]: 👍

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java [hunk 5]
```diff
@@ -119,16 +119,14 @@ public void onFailure(Exception e) {
     }
 
     private Map<String, Object> getMappings(final GetIndexResponse getIndexResponse, final String sourceIndexName) {
-        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = getIndexResponse.mappings();
-        ImmutableOpenMap<String, MappingMetaData> indexMapping = mappings.get(sourceIndexName);
-        if (indexMapping.keys().size() == 0) {
+        ImmutableOpenMap<String, MappingMetaData> mappings = getIndexResponse.mappings();
+        MappingMetaData indexMapping = mappings.get(sourceIndexName);
+        if (indexMapping == null || indexMapping == MappingMetaData.EMPTY_MAPPINGS) {
```
Comment (jtibshirani) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: I'm wondering why we need to check for `null` here?

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java [hunk 6]
```diff
@@ -84,15 +78,21 @@ public GetIndexResponse(String[] indices,
...
-                mappingEntryBuilder.put(in.readString(), new MappingMetaData(in));
+            String index = in.readString();
+            if (in.getVersion().before(Version.V_8_0_0)) {
+                int numMappings = in.readVInt();
+                assert numMappings == 0 || numMappings == 1 : "Expected 0 or 1 mappings but got " + numMappings;
+                if (numMappings == 1) {
+                    String type = in.readString();
+                    assert MapperService.SINGLE_MAPPING_NAME.equals(type) : "Expected [_doc] but got [" + type + "]";
+                    mappingsMapBuilder.put(index, new MappingMetaData(in));
+                }
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: If `numMappings == 0`, I think we would want to put `(index, MappingMetaData.EMPTY_MAPPINGS)`. That way we maintain the invariant that the map contains a key for every requested index.

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java [hunk 7]
```diff
@@ -197,12 +197,19 @@ public String getSetting(String index, String setting) {
...
-        for (ObjectObjectCursor<String, ImmutableOpenMap<String, MappingMetaData>> indexEntry : mappings) {
+        for (ObjectObjectCursor<String, MappingMetaData> indexEntry : mappings) {
             out.writeString(indexEntry.key);
-            out.writeVInt(indexEntry.value.size());
-            for (ObjectObjectCursor<String, MappingMetaData> mappingEntry : indexEntry.value) {
-                out.writeString(mappingEntry.key);
-                mappingEntry.value.writeTo(out);
+            if (out.getVersion().before(Version.V_8_0_0)) {
+                out.writeVInt(indexEntry.value == MappingMetaData.EMPTY_MAPPINGS ? 0 : 1);
+                if (indexEntry.value != null) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Should this be `indexEntry.value != MappingMetaData.EMPTY_MAPPINGS`?

**Review by jtibshirani — COMMENTED** [review]
>  I think I'll open a separate PR to just remove all the type-referring deprecated methods in HLRC to prevent confusion in the future

Seems like a nice idea!

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java (hunk 5 above)
Comment (romseygeek) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: Because the index itself might not exist for some reason, in which case `Map.get(key)` will return `null`

Comment (romseygeek) [reply x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java]: Although looking more closely, the GetIndexRequest used above here will error out if the index doesn't exist, so we can remove the null check.  Good catch!

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java (hunk 6 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: ++

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java (hunk 7 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Missed this one... will fix, thanks!

File: server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java]: True, but then I think we need to throw an exception rather than assert? Or is the idea that assertions are only there in our test infrastructure, and in production it would be ignored.  In which case maybe we should throw an exception here rather than use an assertion?

[the remark this answers is not in the document]
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/action/support/master/info/ClusterInfoRequest.java]: Right, it seems like we should be throwing exceptions here (and in other requests) instead of asserts. This would allow us to 'fail fast' when a user starts a rolling upgrade to 8.x but hasn't fully switched to typeless calls.

I think we use asserts in other requests like `GraphExploreRequest` (I missed that in previous reviews!), perhaps we could update to using exceptions everywhere in a future PR?

**Review by jtibshirani — APPROVED** [review]

**Comment by romseygeek** [comment]
@elasticmachine update branch
